File: Study Plan/Top Interview 150/0012 Integer to Roman/IntegerToRoman.java
/*
    Problem
    --------------------------------------------------------------------------------
    12. Integer to Roman
*/

/*
    Solution
    --------------------------------------------------------------------------------
    - Runtime 7 ms (Beats 39.65% of users with Java)
    - Memory 44.92 MB (Beats 6.19% of users with Java)
*/

class Solution {
    public String intToRoman(int num) {
        StringBuilder roman = new StringBuilder();
        // Guarantees insertion order
        LinkedHashMap<Integer,String> numbersAndSymbols = new LinkedHashMap<>();
        numbersAndSymbols.put(1000, "M" );
        numbersAndSymbols.put( 900, "CM");
        numbersAndSymbols.put( 500, "D" );
        numbersAndSymbols.put( 400, "CD");
        numbersAndSymbols.put( 100, "C" );
        numbersAndSymbols.put(  90, "XC");
        numbersAndSymbols.put(  50, "L" );
        numbersAndSymbols.put(  40, "XL");
        numbersAndSymbols.put(  10, "X" );
        numbersAndSymbols.put(   9, "IX");
        numbersAndSymbols.put(   5, "V" );
        numbersAndSymbols.put(   4, "IV");
        numbersAndSymbols.put(   1, "I" );
        for (Map.Entry<Integer,String> numberAndSymbol : numbersAndSymbols.entrySet()) {
            int number = numberAndSymbol.getKey();
            String symbol = numberAndSymbol.getValue();
            while (num >= number) {
                roman.append(symbol);
                num -= number;
            }
        }
        return roman.toString();
    }
}

File: Study Plan/Top Interview 150/12 Integer to Roman/IntegerToRoman.cs
/*
    Problem
    --------------------------------------------------------------------------------
    12. Integer to Roman
*/

/*
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/integer-to-roman/solutions/4508995/c-time-complexity-o-n-space-complexity-o-1-beats-98-in-runtime-list-of-key-value-pairs/
    
    - Runtime 50 ms (Beats 98.84% of users with C#)
    - Memory 46.38 MB (Beats 26.10% of users with C#)
*/

public class Solution
{
    public string IntToRoman(int num)
    {
        var roman = new StringBuilder();

        // Guarantees insertion order
        List<KeyValuePair<int, string>> numbersAndSymbols = new()
        {
            new(1000, "M" ),
            new( 900, "CM"),
            new( 500, "D" ),
            new( 400, "CD"),
            new( 100, "C" ),
            new(  90, "XC"),
            new(  50, "L" ),
            new(  40, "XL"),
            new(  10, "X" ),
            new(   9, "IX"),
            new(   5, "V" ),
            new(   4, "IV"),
            new(   1, "I" )
        };

        foreach (var numberAndSymbol in numbersAndSymbols)
        {
            var number = numberAndSymbol.Key;
            var symbol = numberAndSymbol.Value;

            while (num >= number)
            {
                roman.Append(symbol);
                num -= number;
            }
        }

        return roman.ToString();
    }
}
